### PR TITLE
pkg/metricutil: fix #516 bad GetCmdLabel() implementation 

### LINKED ISSUE #516: pkg/metricutil: bad GetCmdLabel() implementation
Bad label string!

GetPDMembers => get_p_d_members

ref:  https://github.com/pingcap/pd/blob/master/pkg/metricutil/metricutil.go#L59

```go
// convertName converts variable name to a linux type name.
// Like `AbcDef -> abc_def`.
func convertName(str string) string {
	name := make([]byte, 0, 64)
	for i := 0; i < len(str); i++ {
		if str[i] >= 'A' && str[i] <= 'Z' {
			if i > 0 {
				name = append(name, '_')
			}
			name = append(name, str[i]+'a'-'A')
		} else {
			name = append(name, str[i])
		}
	}
	return string(name)
}
```

### PR DESCRIPTION
**Pull request by andelf**
This fixes #516 .

Also adds some more test cases.

### CODE REVIEW

File: pkg/metricutil/metricutil.go [hunk 1]
```diff
@@ -41,34 +42,32 @@ func GetCmdLabel(request *pdpb.Request) string {
...
-			}
-			name = append(name, str[i]+'a'-'A')
-		} else {
-			name = append(name, str[i])
+func camelCaseToSnakeCase(str string) string {
+	runes := []rune(str)
+	length := len(runes)
+
+	var ret []rune
+	for i := 0; i < length; i++ {
```
Comment (overvenus) [inline pkg/metricutil/metricutil.go]:
```go
for i, re := range str {}
```

**Comment by nolouch** [comment]
LGTM

File: pkg/metricutil/metricutil.go [hunk 2]
```diff
@@ -41,34 +42,32 @@ func GetCmdLabel(request *pdpb.Request) string {
...
+		if i > 0 && unicode.IsUpper(runes[i]) && ((i+1 < length && unicode.IsLower(runes[i+1])) || unicode.IsLower(runes[i-1])) {
```
Comment (siddontang) [inline pkg/metricutil/metricutil.go]: do you format the code ?

**Comment by siddontang** [comment]
PTAL @overvenus

File: pkg/metricutil/metricutil.go (hunk 2 above)
Comment (overvenus) [inline pkg/metricutil/metricutil.go]: Could you simplify this `if`? [Go report card](https://goreportcard.com/report/github.com/pingcap/pd#gocyclo) will blame the complexity.

Comment (andelf) [inline pkg/metricutil/metricutil.go]: This is already the simplest format.

Comment (GregoryIan) [inline pkg/metricutil/metricutil.go]: In accordance with the algebraic, it is indeed the simplest expression. If format, u can split the every preconditions in one ` if line `

Comment (siddontang) [inline pkg/metricutil/metricutil.go]: this line is too long, please split the logic

File: pkg/metricutil/metricutil.go (hunk 1 above)
Comment (andelf) [inline pkg/metricutil/metricutil.go]: meaningless. I need to access previous and next chars in runes. 🤣

**Comment by siddontang** [comment]
no need subject prefix, we can know it directly.

File: pkg/metricutil/metricutil.go [hunk 3]
```diff
@@ -41,34 +42,42 @@ func GetCmdLabel(request *pdpb.Request) string {
...
-	name := make([]byte, 0, 64)
-	for i := 0; i < len(str); i++ {
-		if str[i] >= 'A' && str[i] <= 'Z' {
-			if i > 0 {
-				name = append(name, '_')
-			}
-			name = append(name, str[i]+'a'-'A')
-		} else {
-			name = append(name, str[i])
+func runesHasLowerNeighborAt(runes []rune, idx int) bool {
```
Comment (siddontang) [inline pkg/metricutil/metricutil.go]: use `hasLowerNeighborAt`, no need `runes` here.

**Comment by siddontang** [comment]
PTAL @disksing

File: pkg/metricutil/metricutil.go [hunk 4]
```diff
@@ -41,34 +42,42 @@ func GetCmdLabel(request *pdpb.Request) string {
...
+	if idx > 1 && unicode.IsLower(runes[idx-1]) {
```
Comment (disksing) [inline pkg/metricutil/metricutil.go]: I think you meant `idx > 0`.
_aBCd_ should be converted to _a\_bcd_.

File: pkg/metricutil/metricutil_test.go [hunk 5]
```diff
@@ -45,7 +58,7 @@ func (s *testMetricsSuite) TestConvertName(c *C) {
 	}
 
 	for _, input := range inputs {
-		c.Assert(input.newName, Equals, convertName(input.name))
+		c.Assert(input.newName, Equals, camelCaseToSnakeCase(input.name))
```
Comment (disksing) [inline pkg/metricutil/metricutil_test.go]: Change it to `c.Assert(camelCaseToSnakeCase(input.name), Equals, input.newName)`. Obtained value should be in front of expected value. 
Although it is not introduced by you, thanks for helping to fix it.

**Comment by siddontang** [comment]
Ping @andelf

File: pkg/metricutil/metricutil.go (hunk 4 above)
Comment (andelf) [inline pkg/metricutil/metricutil.go]: a_bc_d

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by disksing** [comment]
LGTM

File: pkg/metricutil/metricutil.go (hunk 3 above)
Comment (disksing) [inline pkg/metricutil/metricutil.go]: @siddontang @andelf Any agreement here? FYI, I'm okay with both..

Comment (siddontang) [inline pkg/metricutil/metricutil.go]: I don't like the style, but if you both think it is ok, then it can be merged.

**Comment by disksing** [comment]
CI failure is not related to this PR and should already be fixed by other PR.
I'm going to merge it.

**Pull request merged**